### ISSUE TITLE
Improve makeValid/isValid further

### ISSUE THREAD

**Issue by hasufell** [haskell/filepath#92]
https://github.com/hasufell/abstract-filepath/issues/10#issuecomment-900782958

Also related: https://github.com/haskell/filepath/issues/56

**Comment by hasufell** [comment]
@Mistuke

**Comment by hasufell** [comment]
I also noticed this, which I find confusing:

```
splitDrive "\\\\?\\UNCshared\\file" == ("\\\\?\\","UNCshared\\file")
```

`"\\\\?\\"` isn't really a drive... is it?

**Comment by Mistuke** [comment]
No, technically speaking that path has no drive, but... Is also invalid.

I'm assuming it was supposed to be `"\\\\?\\UNC\\shared\\file"` since the UNC is required to indicate it's a UNC path.

Not all Paths in Windows have a drive.

> "\\\\?\\" isn't really a drive... is it? 

Indeed, `\\?\` is a marker to tell the win32 api to not process the path but instead send it straight through to the fs driver. 

It's in the so called Win32 “namespace". There are other namespaces like the device on `\\.\`.

Note that because these skip processing from the API layer that you the paths can't contain OS level concept like `.` and `.. `

**Comment by hasufell** [comment]
@Mistuke I just implemented the lexer that Neil was aiming at, which might make a lot of the internals easier and more robust. Please check out my comment here: https://github.com/haskell/filepath/issues/53#issuecomment-985904918

There's likely stuff missing or incorrect, but I think it's a good start.
